Weighted AUROC to omit empty classes

BeyondTheProof opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #348.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

pep8speaks commented:
Hello @BeyondTheProof! Thanks for opening this PR. 

* In the file [`tests/classification/test_auroc.py`](https://github.com/PyTorchLightning/metrics/blob/cfecc54d948800d784cfce92c65e48b546c8edad/tests/classification/test_auroc.py):

> [Line 201:5](https://github.com/PyTorchLightning/metrics/blob/cfecc54d948800d784cfce92c65e48b546c8edad/tests/classification/test_auroc.py#L201): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 212:45](https://github.com/PyTorchLightning/metrics/blob/cfecc54d948800d784cfce92c65e48b546c8edad/tests/classification/test_auroc.py#L212): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 212:103](https://github.com/PyTorchLightning/metrics/blob/cfecc54d948800d784cfce92c65e48b546c8edad/tests/classification/test_auroc.py#L212): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 215:90](https://github.com/PyTorchLightning/metrics/blob/cfecc54d948800d784cfce92c65e48b546c8edad/tests/classification/test_auroc.py#L215): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

* In the file [`torchmetrics/functional/classification/auroc.py`](https://github.com/PyTorchLightning/metrics/blob/cfecc54d948800d784cfce92c65e48b546c8edad/torchmetrics/functional/classification/auroc.py):

> [Line 93:1](https://github.com/PyTorchLightning/metrics/blob/cfecc54d948800d784cfce92c65e48b546c8edad/torchmetrics/functional/classification/auroc.py#L93): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

Borda commented on torchmetrics/functional/classification/auroc.py:
```diff
@@ -89,6 +90,21 @@ def _auroc_compute(
             raise ValueError('Detected input to `multiclass` but you did not provide `num_classes` argument')
+
+        # If one or more classes has 0 observations, we should exclude them, as its weight will be 0 anyway
+        elif average == AverageMethod.WEIGHTED and len(torch.unique(target)) != num_classes:
```
```suggestion
        if average == AverageMethod.WEIGHTED and len(torch.unique(target)) != num_classes:
```
the above ends with `raise`

BeyondTheProof replied:
Good point, will fix!

Borda reviewed (approved):
lgtm

SkafteNicki commented on torchmetrics/functional/classification/auroc.py:
```diff
@@ -89,6 +90,21 @@ def _auroc_compute(
+            class_observed = target_bool_mat.sum(axis=0) > 0
+            for c in range(num_classes):
+                if not class_observed[c]:
+                    warnings.warn(f'Class {c} had 0 observations, omitted from AUROC calculation', UserWarning)
```
could we add a test that checks the warning gets raised for some invalid input?

BeyondTheProof replied:
Added, thank you for the suggestion!

SkafteNicki commented on torchmetrics/functional/classification/auroc.py:
```diff
@@ -89,6 +90,21 @@ def _auroc_compute(
+            target = torch.where(target)[1]
+            num_classes = class_observed.sum()
+            if num_classes <= 1:
+                raise ValueError(f'Found {num_classes} non-empty classes in `multiclass` AUROC calculation')
```
could we add a test that checks the error gets raised for some invalid input?

BeyondTheProof replied:
Done and done!

SkafteNicki reviewed (commented):
Looking good, remember to also update the changelog as this influences the end user!

BeyondTheProof commented:
@SkafteNicki I've addressed all your suggestions, and my tests are passing locally. However, code coverage is failing, but I don't know how to address this, and whether it is important or not in this scenario. I would appreciate this workflow approved to make sure everything is A-OK! Thanks again.

SkafteNicki commented:
@BeyondTheProof do not worry about the code coverage. Only test to focus on are the once marked as "required"

SkafteNicki reviewed (approved):
LGTM :]
